MoltenVK/MoltenVK/GPUObjects/MVKDevice.mm

```diff
@@ -3641,16 +3641,24 @@ static uint32_t mvkGetEntryProperty(io_registry_entry_t entry, CFStringRef prope
 
 MVKBuffer* MVKDevice::getBufferAtAddress(uint64_t address)
 {
-    auto mvkBufIt = _gpuBufferAddressMap.find(address);
+    lock_guard<mutex> lock(_rezLock);
     
-    if(mvkBufIt != _gpuBufferAddressMap.end())
+    std::unordered_map<std::pair<uint64_t, uint64_t>, MVKBuffer*>::iterator it;
```

Comment:
Not particularly pretty (a named struct instead of std::pair would improve readability) and pretty slow but that should work now.

Reply:
Can't we treat the BDA + offset as a buffer? Instead of just storing the ranges?

I mean if it is a "view" of a buffer can't we add it to the unordered_map?

Reply:
> EDIT: I read this as "why would we need to allow for offsets".
> 
> Here's what my toy renderer does for example: 
> Allocate a single huge VkBuffer for all vertex data in the scene and get the BDA of that.
> Then when I build the BLAS, I calculate the VA for where my geometry actually sits which is the BDA + offset.
> That then gets passed to the BLAS build. So if you just look for the base BDA of a buffer, you wouldn't find anything in that case.
> 
> For what to do: store buffer ranges and iterate over them to find one that contains the requested address. To speed it up, you could combine this with a map where you mask off some of the lower bits of the VA as a key.

This buffer is just a single giant buffer, which is then suballocated to then accessed via the offset. I'm not sure how we would go about adding these suballocated buffers as regular buffers.

Reply:
I meant that just to increase the hit rate . If i understood correctly we're always using the base buffer and then always calculating to find the inner buffer.  
This reminds me of Java ByteBuffers and the method slice - https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/ByteBuffer.html#slice() .

Reply:
@rcaridade145 I think you misunderstand the problem.

The application passes a VA (virtual address) and Metal needs a buffer object. It's essentially a raw pointer for the GPU that has to be mapped back to an opaque buffer object.